File: e2e/ergonomic/python/check.py
"""Drives the ergonomic smoke server with real HTTP requests and asserts:

- a valid body  -> 2xx with the typed DTO serialized back
- an invalid body -> 422 ProblemDetails produced by the Rust CORE (not a
  language-side 400), proving validation is delegated to the core.

Exit 0 = pass. Run with the binding-installed interpreter, e.g.
    e2e/python/.venv/bin/python e2e/ergonomic/python/check.py
"""

import json
import os
import subprocess
import sys
import time
import urllib.error
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
PORT = 8000


def post(payload: dict) -> tuple[int, str]:
    req = urllib.request.Request(
        f"http://127.0.0.1:{PORT}/users",
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        resp = urllib.request.urlopen(req, timeout=5)
        return resp.status, resp.read().decode()
    except urllib.error.HTTPError as exc:
        return exc.code, exc.read().decode()


def main() -> int:
    log_path = os.path.join(HERE, ".server.log")
    log = open(log_path, "w")  # noqa: SIM115 (kept open for the child's lifetime)
    proc = subprocess.Popen(
        [sys.executable, os.path.join(HERE, "server.py")],
        stdout=log,
        stderr=subprocess.STDOUT,
    )
    try:
        # Wait for the server to bind (and surface a crash if it exits early).
        for _ in range(60):
            if proc.poll() is not None:
                log.flush()
                with open(log_path) as fh:
                    print(f"FAIL: server exited early rc={proc.returncode}")
                    print(fh.read()[:2500])
                return 1
            try:
                post({"name": "warmup", "email": "w@w.w", "age": 1})
                break
            except OSError:
                time.sleep(0.25)
        else:
            print("FAIL: server never came up")
            return 1

        status, body = post({"name": "Alice", "email": "alice@example.com", "age": 30})
        print(f"VALID   -> {status} {body}")
        if status not in (200, 201) or "Alice" not in body:
            print("FAIL: valid request did not return the typed DTO")
            return 1

        status2, body2 = post({"name": "Bob", "age": "not-a-number"})
        print(f"INVALID -> {status2} {body2}")
        if status2 != 422:
            print(f"FAIL: invalid body expected 422 from the core, got {status2}")
            return 1

        print("ERGO SMOKE PASS (python)")
        return 0
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()


if __name__ == "__main__":
    raise SystemExit(main())

File: e2e/ergonomic/ruby/check.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Drives the ergonomic smoke server with real HTTP requests and asserts:
#
# - a valid body  -> 2xx with the typed DTO serialized back
# - an invalid body -> 422 ProblemDetails produced by the Rust CORE (not a
#   language-side 400), proving validation is delegated to the core.
#
# Exit 0 = pass. Run with the binding-installed Ruby interpreter, e.g.
#     e2e/ergonomic/ruby/check.rb

require "json"
require "net/http"
require "open3"
require "timeout"

HERE = __dir__
PORT = 8000

def post(payload)
  uri = URI("http://127.0.0.1:#{PORT}/users")
  req = Net::HTTP::Post.new(uri)
  req["Content-Type"] = "application/json"
  req.body = JSON.dump(payload)

  http = Net::HTTP.new(uri.host, uri.port)
  http.read_timeout = 5
  resp = http.request(req)
  [resp.code.to_i, resp.body]
rescue StandardError => e
  # Connection failed
  raise e
end

def main
  log_path = File.join(HERE, ".server.log")
  log_file = File.open(log_path, "w")

  # Spawn server in background
  proc = Process.spawn(
    "ruby",
    File.join(HERE, "server.rb"),
    out: log_file,
    err: log_file,
    pgroup: true
  )

  begin
    # Wait for server to bind (with early-exit detection)
    bound = false
    60.times do
      if Process.wait(proc, Process::WNOHANG)
        log_file.flush
        log_content = File.read(log_path)[0, 2500]
        puts("FAIL: server exited early rc=#{$?.exitstatus}")
        puts(log_content)
        return 1
      end

      begin
        post(name: "warmup", age: 1)
        bound = true
        break
      rescue Errno::ECONNREFUSED
        sleep(0.25)
      end
    end

    unless bound
      puts("FAIL: server never came up")
      return 1
    end

    # Test valid request
    status, body = post(name: "Alice", age: 30)
    puts("VALID   -> #{status} #{body}")
    if status < 200 || status >= 300 || !body.include?("Alice")
      puts("FAIL: valid request did not return the typed DTO")
      return 1
    end

    # Test invalid request (age is string, not integer)
    status2, body2 = post(name: "Bob", age: "not-a-number")
    puts("INVALID -> #{status2} #{body2}")
    if status2 != 422
      puts("FAIL: invalid body expected 422 from the core, got #{status2}")
      return 1
    end

    puts("ERGO SMOKE PASS (ruby)")
    0
  ensure
    # Terminate the process group to kill the server and any children.
    begin
      # SIGTERM to the process group
      Process.kill(-15, proc)
      begin
        Timeout.timeout(5) { Process.wait(proc) }
      rescue Timeout::Error
        # SIGKILL if SIGTERM didn't work
        Process.kill(-9, proc)
        Process.wait(proc)
      end

    rescue Errno::ESRCH, Errno::ECHILD
      # Process already gone
    end

    log_file&.close
  end
end

exit(main)
